src/main/cpp/src/map_zip_with_utils.cu

```diff
@@ -0,0 +1,308 @@
+  // Create a column view that represents the keys and values part of col1
+  auto map1_keys = column_view{data_type{type_id::LIST},
+                               col1.size(),
+                               nullptr,
+                               col1.null_mask(),
+                               col1.null_count(),
+                               col1.offset(),
+                               {col1.offsets(), col1.child().child(0)}};
+
+  auto map1_values = column_view{data_type{type_id::LIST},
+                                 col1.size(),
+                                 nullptr,
+                                 col1.null_mask(),
+                                 col1.null_count(),
+                                 col1.offset(),
+                                 {col1.offsets(), col1.child().child(1)}};
+
+  // Extract keys and values from the second map column (col2)
+  // Create a column view that represents the keys and values part of col2
+  auto map2_keys = column_view{data_type{type_id::LIST},
+                               col2.size(),
+                               nullptr,
+                               col2.null_mask(),
+                               col2.null_count(),
+                               col2.offset(),
+                               {col2.offsets(), col2.child().child(0)}};
+
+  auto map2_values = column_view{data_type{type_id::LIST},
+                                 col2.size(),
+                                 nullptr,
+                                 col2.null_mask(),
+                                 col2.null_count(),
+                                 col2.offset(),
+                                 {col2.offsets(), col2.child().child(1)}};
+
+  // Find the union of all unique keys from both maps
+  // This creates a combined set of keys that will be used for the final result
+  auto search_keys      = cudf::lists::union_distinct(cudf::lists_column_view(map1_keys),
+                                                 cudf::lists_column_view(map2_keys));
+  auto search_keys_list = cudf::lists_column_view(*search_keys);
+
+  // Find the indices of each key in the first map
+  // This tells us where each key appears in the first map, or if it doesn't exist
+  auto map1_indices = indices_of(search_keys_list, cudf::lists_column_view(map1_keys), stream, mr);
+  auto map1_indices_list = make_lists_column(search_keys_list.size(),
+                                             std::make_unique<column>(search_keys_list.offsets()),
+                                             std::move(map1_indices),
+                                             0,
+                                             rmm::device_buffer{0, stream, mr});
+
+  // Find the indices of each key in the second map
+  // This tells us where each key appears in the second map, or if it doesn't exist
+  auto map2_indices = indices_of(search_keys_list, cudf::lists_column_view(map2_keys), stream, mr);
```

Comment:
`map2_indices` is a temp column.  We should be using the default mr here.
One option to keep things simple is to use pass the default mr explicitly as the argument to `indices_of()`.